apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/r4/providers/R4SamhsaMatcher.java

```diff
@@ -522,43 +524,95 @@ private static String normalizeIcd10Code(String icd10DiagnosisCode) {
+    if (procedureConcept.getCoding().isEmpty()) {
+      return false;
+    }
+
+    // Does the CodeableConcept have a legit HCPCS Coding?
+    boolean hasHcpcsCoding = hasHcpcsCoding(procedureConcept);
+
+    // Check that Coding to see if it's blacklisted.
+    if (hasHcpcsCoding && isSamhsaCptCode(procedureConcept)) {
+      return true;
+    } else if (!containsOnlyKnownSystems(procedureConcept)) {
+      /*
+       * Fail safe: if we don't know the procedure Coding system, assume the code is
+       * SAMHSA.
+       */
+      return true;
+    } else {
+      // Otherwise, this only has known & non-SAMHSA-blacklisted procedure codes.
+      return false;
+    }
+  }
+
+  /**
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains the
+   *     {@link Coding} with the HCPCS coding system, <code>false</code> if it does not
+   */
+  private boolean hasHcpcsCoding(CodeableConcept procedureConcept) {
     for (Coding procedureCoding : procedureConcept.getCoding()) {
       if (TransformerConstants.CODING_SYSTEM_HCPCS.equals(procedureCoding.getSystem())) {
-        if (isSamhsaCptCode(procedureCoding)) return true;
-      } else {
-        /*
-         * Fail safe: if we don't know the procedure Coding system, assume the code is
-         * SAMHSA.
-         */
         return true;
       }
     }
-
-    // No blacklisted procedure Codings found: this procedure isn't SAMHSA-related.
     return false;
   }
 
   /**
-   * @param procedureCoding the procedure {@link Coding} to check
-   * @return <code>true</code> if the specified procedure {@link Coding} matches one of the {@link
-   *     #cptCodes} entries, <code>false</code> if it does not
+   * @param procedureConcept the procedure {@link CodeableConcept} to check
+   * @return <code>true</code> if the specified procedure {@link CodeableConcept} contains at least
+   *     one {@link Coding} and only contains {@link Coding}s that have known coding systems <code>
+   *     false</code> otherwise
+   */
+  private boolean containsOnlyKnownSystems(CodeableConcept procedureConcept) {
+    Set<String> codingSystems =
+        procedureConcept.getCoding().stream().map(Coding::getSystem).collect(Collectors.toSet());
+
+    String hcpcsCdSystem = CCWUtils.calculateVariableReferenceUrl(CcwCodebookVariable.HCPCS_CD);
+    String hcpcsSystem = TransformerConstants.CODING_SYSTEM_HCPCS;
+
+    // Valid system url for productOrService coding
+    Set<String> knownHcpcsSystem = Set.of(hcpcsSystem);
+
+    // Additional valid coding system URL for backwards-compatibility
+    // See: https://jira.cms.gov/browse/BFD-1345
+    Set<String> backwardsCompatibleHcpcsSystem = Set.of(hcpcsSystem, hcpcsCdSystem);
+
+    return codingSystems.equals(knownHcpcsSystem)
+        || codingSystems.equals(backwardsCompatibleHcpcsSystem);
```

Comment:
It was mentioned I believe that empty collections are handled at a "higher" level, but related to the "defensive coding" ideas mentioned, I think checking for an empty set here could be appropriate.

Reply:
Adding the empty set condition here would make this more permissive and less defensive -- there would be more things that qualify as only having known systems and the failsafe condition is based on whether we have unknowns.

I've gone back and forth a few times on this aspect and ultimately decided to document the more defensive behavior in the method comment here and have an explicit check with comment in the caller.

Reply:
Right, I wasn't meaning in order to make it more defensive, just in the aspect of being thorough.  As you said, the "failsafe" is expected to fire is based on whether we have unknowns, firing "if we don't know the procedure coding system", which doesn't seem as intuitive imo if there is no coding system.
Handling it in the invoking method does of course work, just means you'd have to replicate that step for every method that would want to invoke this one, as it doesn't seem like there'd be a use case, as far as I'm aware of, where you'd want an empty list of codings to actually trigger a samhsa filter.